### PR TITLE
Add a way to opt out of repo code indexing, on a per-repo basis

### PR DESCRIPTION
**Pull request by algernon**
Add a repository setting to opt out of code indexing. When the indexer iterates over repositories, if it sees a repo with the setting disabled, it will remove that repository from the index, otherwise it continues.

The new setting is on by default, and is exposed in the repository settings page, in the advanced settings box, right alongside the Projects and Packages enabling checkboxes.

Addresses #17824, albeit in a different way.

<details>
  <summary>Screenshot</summary>

Screenshot from 2022-07-23 06-49-43

</details>

### CODE REVIEW

**Comment by algernon** [comment]
I turned the setting into something accessible for all repo owners, rather than limiting it to admins. The only part I'm unsure about is where toggling the option on or off and updating the settings, `code.UpdateRepoIndexer()` is triggered. That's something another option hidden behind an admin-only flag does too, so perhaps it would make more sense to not call it here, and just have the next indexer run - whenever it happens - take care of updating things. However, that has the downside of the change potentially taking a long time to take effect.

**Comment by lafriks** [comment]
I would say that changing option should trigger repository reindex as it may not happen ever if repository is not changed

File: models/repo/repo.go [hunk 1]
```diff
@@ -158,6 +158,7 @@ type Repository struct {
 	CodeIndexerStatus               *RepoIndexerStatus `xorm:"-"`
 	StatsIndexerStatus              *RepoIndexerStatus `xorm:"-"`
 	IsFsckEnabled                   bool               `xorm:"NOT NULL DEFAULT true"`
+	IsCodeIndexingEnabled           bool               `xorm:"NOT NULL DEFAULT true"`
```
Comment (Gusted) [inline models/repo/repo.go:161]: This require a migration, it should be similar to [`models/migrations/v219.go`](https://github.com/go-gitea/gitea/blob/main/models/migrations/v219.go), copy paste this whole struct(fixup the imports) and run Sync2 on it.

File: templates/repo/settings/options.tmpl [hunk 2]
```diff
@@ -420,6 +420,16 @@
 					</div>
 				</div>
 
+				{{if .CodeIndexerEnabled}}
+					<div class="inline field">
+						<label>{{.locale.Tr "repo.settings.code_search_indexing"}}</label>
+						<div class="ui checkbox">
+							<input class="enable-system" name="enable_code_indexing" type="checkbox" {{if .Repository.IsCodeIndexingEnabled}}checked{{end}}>
+							<label>{{.locale.Tr "repo.settings.code_search_indexing_desc"}}</label>
```
Comment (delvh) [inline templates/repo/settings/options.tmpl:428]:
```suggestion
							<input type="checkbox" class="enable-system" id="code-indexing-checkbox" name="enable_code_indexing"{{if .Repository.IsCodeIndexingEnabled}} checked{{end}}>
							<label for="code-indexing-checkbox">{{.locale.Tr "repo.settings.code_search_indexing_desc"}}</label>
```

File: routers/web/repo/setting.go [hunk 3]
```diff
@@ -394,6 +395,11 @@ func SettingsPost(ctx *context.Context) {
 			repoChanged = true
 		}
 
+		if repo.IsCodeIndexingEnabled != form.EnableCodeIndexing {
+			repo.IsCodeIndexingEnabled = form.EnableCodeIndexing
+			codeIndexingChanged = true
+		}
```
Comment (delvh) [inline routers/web/repo/setting.go:401]:
```suggestion
		codeIndexingChanged = repo.IsCodeIndexingEnabled != form.EnableCodeIndexing
		if codeIndexingChanged {
			repo.IsCodeIndexingEnabled = form.EnableCodeIndexing
		}
```

File: models/repo/repo.go (hunk 1 above)
Comment (algernon) [reply models/repo/repo.go:161]: Will do!

May take a day or two until I have some cycles to spend on updating this PR, but I'll get there eventually.

**Pull request closed without merging**